File: src/types/message.ts
/**
 * Core message types for LLM conversations
 * 
 * Base types used across the system for message handling.
 * These are intentionally minimal and can be extended in hawk.
 */

import type { ContentBlockParam } from './sdk.js'

// ============================================================================
// Content Blocks
// ============================================================================

/**
 * Text content block
 */
export interface TextBlock {
  type: 'text'
  text: string
}

/**
 * Image content block
 */
export interface ImageBlock {
  type: 'image'
  source: {
    type: 'base64'
    media_type: 'image/jpeg' | 'image/png' | 'image/gif' | 'image/webp'
    data: string
  }
}

/**
 * Tool use block - when assistant uses a tool
 */
export interface ToolUseBlock {
  type: 'tool_use'
  id: string
  name: string
  input: Record<string, unknown>
}

/**
 * Tool result block - result from tool execution
 */
export interface ToolResultBlock {
  type: 'tool_result'
  tool_use_id: string
  content: string | Array<TextBlock | ImageBlock>
  is_error?: boolean
}

/**
 * Union type for all content blocks
 */
export type ContentBlock = TextBlock | ImageBlock | ToolUseBlock | ToolResultBlock

// ============================================================================
// Base Message Types
// ============================================================================

/**
 * Base message interface
 */
export interface Message {
  role: 'user' | 'assistant' | 'system'
  content: string | ContentBlock[]
}

/**
 * User message
 */
export interface UserMessage extends Message {
  role: 'user'
}

/**
 * Assistant message
 */
export interface AssistantMessage extends Message {
  role: 'assistant'
  /**
   * Indicates this message represents an API error
   */
  isApiErrorMessage?: boolean
  /**
   * Error details if this is an error message
   */
  errorDetails?: {
    actualTokens?: number
    limitTokens?: number
  }
}

/**
 * System message
 */
export interface SystemMessage extends Message {
  role: 'system'
}

// ============================================================================
// Message Origins & Sources
// ============================================================================

/**
 * Where a message originated from
 */
export type MessageOrigin = 
  | 'user' 
  | 'api' 
  | 'tool' 
  | 'system' 
  | 'compact'
  | 'recovery'

/**
 * Source of the message (for mailbox/pipeline)
 */
export type MessageSource = 
  | 'user' 
  | 'teammate' 
  | 'system' 
  | 'tick' 
  | 'task'

// ============================================================================
// Utility Types
// ============================================================================

/**
 * Type guard for text blocks
 */
export function isTextBlock(block: unknown): block is TextBlock {
  return (
    typeof block === 'object' &&
    block !== null &&
    'type' in block &&
    block.type === 'text' &&
    'text' in block &&
    typeof block.text === 'string'
  )
}

/**
 * Type guard for image blocks
 */
export function isImageBlock(block: unknown): block is ImageBlock {
  return (
    typeof block === 'object' &&
    block !== null &&
    'type' in block &&
    block.type === 'image' &&
    'source' in block
  )
}

/**
 * Type guard for tool use blocks
 */
export function isToolUseBlock(block: unknown): block is ToolUseBlock {
  return (
    typeof block === 'object' &&
    block !== null &&
    'type' in block &&
    block.type === 'tool_use' &&
    'id' in block &&
    'name' in block
  )
}

/**
 * Type guard for tool result blocks
 */
export function isToolResultBlock(block: unknown): block is ToolResultBlock {
  return (
    typeof block === 'object' &&
    block !== null &&
    'type' in block &&
    block.type === 'tool_result' &&
    'tool_use_id' in block
  )
}

// ============================================================================
// Conversion Utilities
// ============================================================================

/**
 * Convert content blocks to SDK format
 */
export function toSDKContentBlocks(
  blocks: ContentBlock[],
): ContentBlockParam[] {
  return blocks.map(block => {
    if (block.type === 'text') {
      return { type: 'text', text: block.text }
    }
    if (block.type === 'image') {
      return {
        type: 'image',
        source: block.source,
      }
    }
    if (block.type === 'tool_use') {
      return {
        type: 'tool_use',
        id: block.id,
        name: block.name,
        input: block.input,
      }
    }
    if (block.type === 'tool_result') {
      return {
        type: 'tool_result',
        tool_use_id: block.tool_use_id,
        content: block.content,
        is_error: block.is_error,
      }
    }
    throw new Error(`Unknown block type: ${(block as { type: string }).type}`)
  })
}

/**
 * Create a user message with text content
 */
export function createUserMessage(text: string): UserMessage {
  return {
    role: 'user',
    content: text,
  }
}

/**
 * Create an assistant message with text content
 */
export function createAssistantMessage(text: string): AssistantMessage {
  return {
    role: 'assistant',
    content: text,
  }
}

/**
 * Create a system message
 */
export function createSystemMessage(text: string): SystemMessage {
  return {
    role: 'system',
    content: text,
  }
}

File: types/message.go
package types

// --- SDK content block types ---

type TextBlock struct {
	Type string `json:"type"`
	Text string `json:"text"`
}

type TextBlockParam struct {
	Type string `json:"type"`
	Text string `json:"text"`
}

type Base64ImageSource struct {
	Type      string `json:"type"`
	MediaType string `json:"media_type"`
	Data      string `json:"data"`
}

type ImageBlockParam struct {
	Type   string            `json:"type"`
	Source Base64ImageSource `json:"source"`
}

type ToolUseBlock struct {
	Type  string                 `json:"type"`
	ID    string                 `json:"id"`
	Name  string                 `json:"name"`
	Input map[string]interface{} `json:"input"`
}

type ToolUseBlockParam struct {
	Type  string                 `json:"type"`
	ID    string                 `json:"id"`
	Name  string                 `json:"name"`
	Input map[string]interface{} `json:"input"`
}

type ToolResultBlockParam struct {
	Type      string      `json:"type"`
	ToolUseID string      `json:"tool_use_id"`
	Content   interface{} `json:"content"`
	IsError   bool        `json:"is_error"`
}

type ThinkingBlock struct {
	Type      string `json:"type"`
	Thinking  string `json:"thinking"`
	Signature string `json:"signature"`
}

type ThinkingBlockParam struct {
	Type      string `json:"type"`
	Thinking  string `json:"thinking"`
	Signature string `json:"signature"`
}

type RedactedThinkingBlock struct {
	Type string `json:"type"`
	Data string `json:"data"`
}

type RedactedThinkingBlockParam struct {
	Type string `json:"type"`
	Data string `json:"data"`
}

type (
	ContentBlock      = interface{}
	ContentBlockParam = interface{}
)

// --- Message types ---

type Usage struct {
	InputTokens  int `json:"input_tokens"`
	OutputTokens int `json:"output_tokens"`
}

type Message struct {
	ID           string      `json:"id"`
	Type_        string      `json:"type"`
	Role         string      `json:"role"`
	Content      interface{} `json:"content"`
	Model        string      `json:"model"`
	StopReason   string      `json:"stop_reason"`
	StopSequence string      `json:"stop_sequence"`
	Usage        *Usage      `json:"usage"`
}

type MessageParam struct {
	Role    string      `json:"role"`
	Content interface{} `json:"content"`
}

type BetaUsage struct {
	InputTokens  int `json:"input_tokens"`
	OutputTokens int `json:"output_tokens"`
}

type BetaMessage struct {
	ID           string         `json:"id"`
	Type_        string         `json:"type"`
	Role         string         `json:"role"`
	Content      []ContentBlock `json:"content"`
	Model        string         `json:"model"`
	StopReason   string         `json:"stop_reason"`
	StopSequence string         `json:"stop_sequence"`
	Usage        BetaUsage      `json:"usage"`
}

type BetaMessageParam struct {
	Role    string      `json:"role"`
	Content interface{} `json:"content"`
}

type ToolInputSchema struct {
	Type       string                 `json:"type"`
	Properties map[string]interface{} `json:"properties"`
	Required   []string               `json:"required"`
}

type Tool struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	InputSchema ToolInputSchema `json:"input_schema"`
}

// --- StopReason constants ---

const (
	EndTurn      = "end_turn"
	MaxTokens    = "max_tokens"
	StopSequence = "stop_sequence"
	ToolUse      = "tool_use"
)

// --- MessageOrigin constants ---

type MessageOrigin = string

const (
	User       MessageOrigin = "user"
	API        MessageOrigin = "api"
	ToolOrigin MessageOrigin = "tool"
	System     MessageOrigin = "system"
	Compact    MessageOrigin = "compact"
	Recovery   MessageOrigin = "recovery"
)

// --- MessageSource constants ---

type MessageSource = string

const (
	UserSource   MessageSource = "user"
	Teammate     MessageSource = "teammate"
	SystemSource MessageSource = "system"
	Tick         MessageSource = "tick"
	Task         MessageSource = "task"
)

// --- Specialized message types ---

type ErrorDetails struct {
	ActualTokens *int `json:"actual_tokens,omitempty"`
	LimitTokens  *int `json:"limit_tokens,omitempty"`
}

type UserMessage struct {
	Message
}

type AssistantMessage struct {
	Message
	IsApiErrorMessage bool          `json:"is_api_error_message"`
	ErrorDetails      *ErrorDetails `json:"error_details,omitempty"`
}

type SystemMessage struct {
	Message
}

// --- Request/config types ---

type ToolChoice struct {
	Type string `json:"type"`
	Name string `json:"name,omitempty"`
}

type MessageCreateParams struct {
	Model         string         `json:"model"`
	MaxTokens     int            `json:"max_tokens"`
	Messages      []MessageParam `json:"messages"`
	Tools         []Tool         `json:"tools,omitempty"`
	ToolChoice    *ToolChoice    `json:"tool_choice,omitempty"`
	System        *string        `json:"system,omitempty"`
	Temperature   *float64       `json:"temperature,omitempty"`
	TopP          *float64       `json:"top_p,omitempty"`
	TopK          *int           `json:"top_k,omitempty"`
	StopSequences []string       `json:"stop_sequences,omitempty"`
	Stream        bool           `json:"stream"`
}

type MessageStreamEvent struct {
	Type string `json:"type"`
}

type ClientOptions struct {
	APIKey     string `json:"-"`
	BaseURL    string `json:"base_url"`
	Timeout    int    `json:"timeout"`
	MaxRetries int    `json:"max_retries"`
}

// --- Type guard functions ---

func IsTextBlock(v interface{}) (TextBlock, bool) {
	b, ok := v.(TextBlock)
	return b, ok
}

func IsImageBlock(v interface{}) (ImageBlockParam, bool) {
	b, ok := v.(ImageBlockParam)
	return b, ok
}

func IsToolUseBlock(v interface{}) (ToolUseBlock, bool) {
	b, ok := v.(ToolUseBlock)
	return b, ok
}

func IsToolResultBlock(v interface{}) (ToolResultBlockParam, bool) {
	b, ok := v.(ToolResultBlockParam)
	return b, ok
}

// --- Creator functions ---

func CreateUserMessage(text string) UserMessage {
	return UserMessage{
		Message: Message{
			Role:    "user",
			Content: text,
		},
	}
}

func CreateAssistantMessage(text string) AssistantMessage {
	return AssistantMessage{
		Message: Message{
			Role:    "assistant",
			Content: text,
		},
	}
}

func CreateSystemMessage(text string) SystemMessage {
	return SystemMessage{
		Message: Message{
			Role:    "system",
			Content: text,
		},
	}
}
